Source/deviceidentification/DeviceIdentification.cpp

```diff
@@ -0,0 +1,269 @@
+            }
+
+            _adminLock.Unlock();
+
+            return result;
+        }
+
+    private:
+        DeviceIdentification* Find(const string& name)
+        {
+            DeviceIdentification* result(nullptr);
+
+            std::list<DeviceIdentification*>::iterator index(std::list<DeviceIdentification*>::begin());
+
+            while ((index != std::list<DeviceIdentification*>::end()) && ((*index)->Name() != name)) {
+                index++;
+            }
+
+            if (index != std::list<DeviceIdentification*>::end()) {
+                result = *index;
+                result->AddRef();
+            }
+
+            return result;
+        }
+
+        Core::CriticalSection _adminLock;
+        Core::ProxyType<RPC::InvokeServerType<1, 0, 8>> _engine;
+        Core::ProxyType<RPC::CommunicatorClient> _comChannel;
+    };
+
+    mutable int _refCount;
+    const string _name;
+    Exchange::IDeviceProperties* _deviceConnection;
+    PluginHost::ISubSystem::IIdentifier* _identifier;
+    static DeviceIdentification::DeviceIdAdministration _administration;
+
+public:
+    DeviceIdentification() = delete;
+    DeviceIdentification(const DeviceIdentification&) = delete;
+    DeviceIdentification& operator=(const DeviceIdentification&) = delete;
+
+    static DeviceIdentification* Instance(const string& name)
+    {
+        return _administration.Instance(name);
+    }
+    void AddRef() const
+    {
+        Core::InterlockedIncrement(_refCount);
+    }
+    uint32_t Release() const
+    {
+        return _administration.Delete(this, _refCount);
+    }
+
+    const string& Name() const { return _name; }
+
+    int16_t Chipset(char buffer[], const uint8_t length) const
+    {
+        int16_t stringLength = 0;
+
+        if (_deviceConnection != nullptr) {
+            string chipset = _deviceConnection->Chipset();
+            stringLength = chipset.length();
+
+            if (length > stringLength) {
+                strncpy(buffer, chipset.c_str(), length);
+            } else {
+                return -stringLength;
+            }
+        }
+        return stringLength;
+    }
+
+    int16_t FirmwareVersion(char buffer[], const uint8_t length) const
+    {
+        int16_t stringLength = 0;
+
+        if (_deviceConnection != nullptr) {
+            string firmware = _deviceConnection->FirmwareVersion();
+            stringLength = firmware.length();
+
+            if (length > stringLength) {
+                strncpy(buffer, firmware.c_str(), length);
+            } else {
+                return -stringLength;
+            }
+        }
+        return stringLength;
+    }
+
+    int16_t Identifier(uint8_t buffer[], const uint8_t length) const
+    {
+        int16_t result = 0;
+
+        if (_identifier != nullptr) {
+
+            result = _identifier->Identifier(length, buffer);
+
+            if (length <= result) {
+
+                return -result;
+            }
```

Comment:
I think this is incorrect. Identifier() will return the number of bytes it has written, so you cannot tell this way what is the required output. Also "length" will never be less than "result". To iron this out and give the client a coherent interface I guess in case (length == result) you'll need to call Identifier() again with an oversized buffer for it to tell you the correct size...?